// Copyright 2022 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package xls.verilog;

// Metrics collected for the block after block conversion completes.
message BlockMetricsProto {
  // The total number of registers (in bits) in the block.
  optional int64 flop_count = 1;

  // Whether the block has a combinational path from an input port to an output
  // port.
  optional bool feedthrough_path_exists = 2;

  // The name of the delay model used to generate the block.
  optional string delay_model = 3;

  // The maximum combinational delay in picoseconds of any path in the block
  // from the output of a register to the input of a register.
  optional int64 max_reg_to_reg_delay_ps = 4;

  // The maximum combinational delay in picoseconds of any path in the block
  // from an input port to a register.
  optional int64 max_input_to_reg_delay_ps = 5;

  // The maximum combinational delay in picoseconds of any path in the block
  // from a register to an output port.
  optional int64 max_reg_to_output_delay_ps = 6;

  // The maximum delay in picoseconds of any combinational path from an input
  // port to an output port.
  optional int64 max_feedthrough_path_delay_ps = 7;
}

message XlsMetricsProto {
  optional BlockMetricsProto block_metrics = 1;
}
